baseplate/experiments/providers/legacy.py

```diff
@@ -0,0 +1,233 @@
+        # while keeping the experience consistent for users in any group other
+        # than excluded users:
+        #
+        #     A B C A B C A B C A B C A B C A B x A B x x x x...
+        #
+        # Rather than building this entire structure out in memory, we can use
+        # a little bit of math to figure out just the one bucket's value.
+        num_variants = len(self.variants)
+        variant_names = sorted(self.variants.keys())
+        # If the variants took up the entire set of buckets, which bucket would
+        # we be in?
+        candidate_variant = variant_names[bucket % num_variants]
+        # Log a warning if this variant is capped, to help us prevent user (us)
+        # error.  It's not the most correct to only check the one, but it's
+        # easy and quick, and anything with that high a percentage should be
+        # selected quite often.
+        variant_fraction = self.variants[candidate_variant] / 100.0
+        variant_cap = 1.0 / num_variants
+        if variant_fraction > variant_cap:
+            logger.warning(
+                'Variant %s exceeds allowable percentage (%.2f > %.2f)',
+                candidate_variant,
+                variant_fraction,
+                variant_cap,
+            )
+        # Variant percentages are expressed as numeric percentages rather than
+        # a fraction of 1 (that is, 1.5 means 1.5%, not 150%); thus, at 100
+        # buckets, buckets and percents map 1:1 with each other.  Since we may
+        # have more than 100 buckets (causing each bucket to represent less
+        # than 1% each), we need to scale up how far "right" we move for each
+        # variant percent.
+        bucket_multiplier = self.num_buckets / 100
+        # Now check to see if we're far enough left to be included in the
+        # variant percentage.
+        bucket_limit = (
+            self.variants[candidate_variant] *
+            num_variants *
+            bucket_multiplier
+        )
+        if bucket < bucket_limit:
+            return candidate_variant
+        else:
+            return None
+
+    def _get_page_experiment_variant(self, content_id, content_type):
+        bucket = self._get_thing_bucket(content_id, content_type)
+        if bucket is None:
+            return None
+        return self._choose_variant(bucket)
+
+    def _get_user_experiment_variant(self, user_id):
+        if user_id is None:
+            return None
+
+        bucket = self._calculate_bucket(user_id)
+        return self._choose_variant(bucket)
+
+    def _get_thing_bucket(self, content_id, content_type):
+        if content_id is None:
+            return None
+
+        # If we've restricted the experiment to certain page types, make sure
+        # the request is for one of those
+        if (self.content_flags.get('subreddit_only', False) and
+                content_type != 'subreddit'):
+            return None
+
+        if (self.content_flags.get('link_only', False) and
```

Comment:
just to play devil's advocate, if we add a completely new content type we'd have to update this library too?